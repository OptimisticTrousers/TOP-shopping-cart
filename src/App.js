import React from "react";
import { Outlet } from "react-router-dom";
import Navbar from "./components/Navbar";
import Footer from "./components/Footer";
import { CartProvider } from "./context/Store";

export default function App() {

  return (
    <div className="container">
      <CartProvider>
        <Navbar />
        <main>
          <Outlet />
          <Footer />
        </main>
      </CartProvider>
    </div>
  );
}