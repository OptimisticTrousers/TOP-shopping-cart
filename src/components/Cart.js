import CartItem from "./CartItem"
import logo from '../assets/optimistictrousers.jpg'

export default function Cart() {

    return (
        <div className="cart-container">
            <h1 className="cart-title">Your Cart</h1>
            <div className="cart">
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>

            </div>
            <p>SUBTOTAL</p>
            <div className="buttons">
                <button type="button">Check Out</button>
                <button type="button">Back To Products</button>
            </div>
        </div>
    )
}