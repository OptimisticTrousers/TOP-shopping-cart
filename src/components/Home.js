import Hero from "./Hero"
import About from "./About"

export default function Home() {
    return (
        <div className="home">
            <Hero /> 
            <About />
        </div>
    )
}