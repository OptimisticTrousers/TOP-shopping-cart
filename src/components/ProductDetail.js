import { Link, useParams } from "react-router-dom"
import { useEffect, useState } from "react"
import leftArrow from '../assets/left-arrow.svg'
import { useAddToCart } from "../context/Store"

export default function ProductDetail(){

    const addToCart = useAddToCart()

    const params = useParams()

    const [details, setDetails] = useState({})

    useEffect(() => {
        fetch(`https://fakestoreapi.com/products/${params.id}`)
            .then(res => res.json())
            .then(data => setDetails(data))
    }, [])

    return (
        <div className="detail-container">
            <div className="image-detail-view">
                <img className="image-detail" src={details.image} alt={details.title}/>
            </div>
            <div className="product-details">
                <Link to="/shop">
                    <button type="button"><img className="left-arrow" src={leftArrow} alt="left arrow" /> Back To Products</button>
                </Link>
                <h1 className="product-title">{details.title}</h1>
                <p className="product-price">${details.price}</p>
                <p className="category-detail">Category: {details.category}</p>
                <p className="description-detail">{details.description}</p>
                <button onClick={() => addToCart({...details, quantity: 0})} className="add-to-cart-button" type="button">Add to Cart</button>
            </div>
        </div>
    )
}