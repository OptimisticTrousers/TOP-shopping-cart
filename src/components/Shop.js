export default function Shop() {
    return (
        <main>
            <h1>Get Clothing!</h1>
            <p>Times are tough. Liven up with our top-notch clothing!</p>
            <div className="products">

            </div>
        </main>
    )
}