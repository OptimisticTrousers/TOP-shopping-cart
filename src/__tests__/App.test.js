    
    import React from 'react'
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import App from '../App'
import Home from '../components/Home'
import Shop from '../components/Shop'
import ProductDetail from '../components/ProductDetail'
import Cart from '../components/Cart'

describe("App" , () => {

    beforeEach(() => {

        fetch.mockClear()

        fetch.mockResponseOnce(JSON.stringify([{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }]))
    })

    it('user can increment quantity', async () => {

        render(
            <React.StrictMode>
                <BrowserRouter>
                    <Routes>
                        <Route path="/" element={<App />} >
                            <Route index element={<Home />} />
                            <Route path="shop" element={<Shop />} />
                            <Route path="shop/:id" element={<ProductDetail />} />
                            <Route path="cart" element={<Cart />} />
                        </Route>
                    </Routes>
                </BrowserRouter>
            </React.StrictMode>
        )

        const user = userEvent.setup()

        // Adding first product to cart
        await user.click(screen.queryByText("SHOP"))
        const firstProduct = await screen.findByAltText('Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops')
        await user.click(firstProduct)
        await user.click(screen.getByRole('button', {name: /Add to Cart/i}))

        // Adding second product to cart
        await user.click(screen.queryByText("SHOP"))
        const secondProduct = await screen.findByAltText('Mens Casual Premium Slim Fit T-Shirts')
        await user.click(secondProduct)
        await user.click(screen.getByRole('button', {name: /Add to Cart/i}))

        // Going to checkout
        const cartSVG = screen.getByTestId('cart-svg')
        await user.click(cartSVG)

        //Clicking quantity incrementor for first product only
        const [firstItemTitle] = await screen.findAllByRole('button', {name: "+"})
        await user.click(firstItemTitle)
        const firstItemQuantity = await screen.findAllByTestId('product-quantity')


        expect(firstItemQuantity).toBe("2")
    })
})
