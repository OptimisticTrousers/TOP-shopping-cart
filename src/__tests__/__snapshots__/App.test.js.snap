// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav>
      <div
        class="links"
      >
        <div
          class="logo"
        >
          <a
            href="/"
          >
            <img
              alt="smiling pair of pants"
              class="logo-image"
              src="optimistictrousers.jpg"
            />
          </a>
          <h1
            class="nav-title"
          >
            The Optimistic Store
          </h1>
          <ul>
            <li>
              <a
                href="/"
              >
                <h2>
                  HOME
                </h2>
              </a>
            </li>
            <li>
              <a
                href="/shop"
              >
                <h2>
                  SHOP
                </h2>
              </a>
            </li>
          </ul>
        </div>
        <a
          href="/cart"
        >
          <span
            data-testid="quantity"
          >
            0
          </span>
          <svg
            data-testid="cart-svg"
            fill="none"
            height="22"
            stroke="currentColor"
            viewBox="0 0 20 22"
            width="20"
          >
            <path
              d="M4 1L1 5V19C1 19.5304 1.21071 20.0391 1.58579 20.4142C1.96086 20.7893 2.46957 21 3 21H17C17.5304 21 18.0391 20.7893 18.4142 20.4142C18.7893 20.0391 19 19.5304 19 19V5L16 1H4Z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M1 5H19"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M14 9C14 10.0609 13.5786 11.0783 12.8284 11.8284C12.0783 12.5786 11.0609 13 10 13C8.93913 13 7.92172 12.5786 7.17157 11.8284C6.42143 11.0783 6 10.0609 6 9"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
        </a>
      </div>
    </nav>
    <main>
      <div
        class="home"
      >
        <div
          class="hero"
        >
          <h1>
            The Optimistic Store
          </h1>
        </div>
        <div
          class="about"
        />
      </div>
    </main>
    <footer />
  </div>
</DocumentFragment>
`;
